internal/memory/pool.go

```diff
@@ -0,0 +1,323 @@
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+// Allocate checks memory pool of class `memCls` and returns a free region with minimal offset, if none available,
+// tries to expanding current `memCls` pool.
+//
+// Internally it's done via moving a region from free pool into a busy pool
+func (pa *PoolAllocator) Allocate(memCls classType) (MappedRegion, error) {
+	if memCls >= memoryClassNumber {
+		return nil, ErrInvalidMemoryClass
+	}
+
+	// Find smallest available offset
+	offset, err := pa.FindFreeOffset(memCls)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mark region at a given offset as busy
+	if r, err := pa.markBusy(memCls, offset); err != nil {
+		if err := pa.Expand(memCls); err != nil {
+			if err == ErrInvalidMemoryClass {
+				return nil, ErrNotEnoughSpace
+			}
+			return nil, err
+		}
+	} else {
+		return r, nil
+	}
+
+	// Now the region at `offset` location should be available
+	if r, _ := pa.markBusy(memCls, offset); r != nil {
+		return r, nil
+	}
+
+	return nil, ErrNotEnoughSpace
+}
+
+// Expand tries to recursively split a bigger memory region into smaller ones until it succeeds or hits the upper limit
+func (pa *PoolAllocator) Expand(clsType classType) error {
+	nextClsType := clsType + 1
+	if nextClsType >= memoryClassNumber {
+		return ErrInvalidMemoryClass
+	}
+
+	nextPool := pa.pools[nextClsType]
+	if nextPool == nil {
+		nextPool = newEmptyMemoryPool()
+		pa.pools[nextClsType] = nextPool
+	}
+
+	// not enough memory in the next class, try to recursively expand
+	if len(nextPool.free) == 0 {
+		err := pa.Expand(nextClsType)
+		if err != nil {
+			return err
+		}
+	}
+
+	offset, err := pa.FindFreeOffset(nextClsType)
+	if err != nil {
+		return err
+	}
+
+	nextReg, err := pa.markBusy(nextClsType, offset)
+	if err != nil {
+		return err
+	}
+
+	// memCls validity has been checked already, we can ignore the error
+	clsSize, _ := GetMemoryClassSize(clsType)
+
+	// expand memCls
+	cp := pa.pools[clsType]
+	if cp == nil {
+		cp = newEmptyMemoryPool()
+		pa.pools[clsType] = cp
+	}
+	// create 4 smaller regions
+	for i := uint64(0); i < 4; i++ {
+		offset := nextReg.offset + i*clsSize
+		reg := &region{
+			parent: nextReg,
+			class:  clsType,
+			offset: offset,
+		}
+		cp.free[offset] = reg
+	}
+	return nil
+}
+
+// Release marks a memory region of class `memCls` and offset `offset` as free and tries to merge smaller regions into
+// a bigger one
+func (pa *PoolAllocator) Release(memCls classType, offset uint64) error {
+	mp := pa.pools[memCls]
+	if mp == nil {
+		return ErrEmptyPoolOperation
+	}
+
+	n, err := pa.markFree(memCls, offset)
+	if err != nil {
+		return err
+	}
+
+	if err := pa.merge(n.parent); err != nil {
+		if err != ErrEarlyMerge {
+			return err
+		}
+	}
+	return nil
+}
+
+func (pa *PoolAllocator) merge(parent *region) error {
+	// nothing to merge
+	if parent == nil {
+		return nil
+	}
+
+	parentCls := parent.class
+	// nothing to do, cannot merge the smallest class
+	if parentCls == 0 {
+		return nil
+	}
```

Comment:
Will we ever have `parentCls` as 0? For that to happen we must have already split the 0'th class right?